Sources/NIOPosix/System.swift

```diff
@@ -644,12 +740,91 @@ internal enum Posix {
     }
 
     @inline(never)
-    internal static func unlink(pathname: String) throws {
+    internal static func lstat(pathname: String, outStat: UnsafeMutablePointer<stat>) throws {
+        _ = try syscall(blocking: false) {
+            sysLstat(pathname, outStat)
+        }
+    }
+
+    @inline(never)
+    public static func symlink(pathname: String, destination: String) throws {
+        _ = try syscall(blocking: false) {
+            sysSymlink(destination, pathname)
+        }
+    }
+
+    @inline(never)
+    public static func readlink(pathname: String, outPath: UnsafeMutablePointer<CChar>, outPathSize: Int) throws -> CLong {
+        return try syscall(blocking: false) {
+            sysReadlink(pathname, outPath, outPathSize)
+        }.result
+    }
+
+    @inline(never)
+    public static func unlink(pathname: String) throws {
         _ = try syscall(blocking: false) {
             sysUnlink(pathname)
         }
     }
 
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    @inline(never)
+    public static func opendir(pathname: String) throws -> UnsafeMutablePointer<DIR> {
+        return try syscall {
+            sysOpendir(pathname)
+        }
+    }
+
+    @inline(never)
+    public static func readdir(dir: UnsafeMutablePointer<DIR>) throws -> UnsafeMutablePointer<dirent>? {
+        return try syscallOptional {
+            sysReaddir(dir)
```

Comment:
Can we set `errno = 0` before this call?

Reply:
added this to `syscallOptional `